src/azure-cli/azure/cli/command_modules/util/custom.py

```diff
@@ -187,10 +188,12 @@ def _upgrade_on_windows():
     """
     logger.warning("Updating Azure CLI with MSI from https://aka.ms/installazurecliwindows")
     tmp_dir, msi_path = _download_from_url('https://aka.ms/installazurecliwindows')
-
     logger.warning("Installing MSI")
     import subprocess
-    exit_code = subprocess.call(['msiexec.exe', '/i', msi_path])
+    subprocess.Popen(['msiexec.exe', '/i', msi_path])
+    logger.warning("Installation started, please wait for a few minutes.")
+    import sys
+    sys.exit(0)
```

Comment:
As stated in the PR description, this makes us be unable to tell if the upgrade is successful. Better to confirm with PM whether this is acceptable solution. Otherwise, we can develop some "upgrader.exe" written in C.

Reply:
We can create a upgrader to update az to prevent running az, but we need to call `az` to map `az upgrade` command to that upgrader.

`az upgrade` is very complicated. It downloads MSI, waits for installation to finish and updates extensions in one command.
I can hardly come up with an idea to fix this issue without breaking change.

Reply:
I thought about it again and feel it is indeed not possible. A GUI application can have such `upgrader.exe` to upgrade the product itself, but this can't be done for a console application.

Consider this invocation chain: `python.exe`(1) -> `upgrader.exe` -> `python.exe`(2)

- To keep the second `python.exe` connected to the original terminal context, every executable on the invocation chain must be alive and wait for its downstream executable to exit. In order words, if `python.exe`(1) subprocesses `upgrader.exe` and exits, then after `upgrader.exe` upgrades `python.exe`, subprocesses `python.exe`(2) and exits, it is no longer possible to connect `python.exe`(2) back to the original terminal context - the output is out of order.
- If the first `python.exe` is alive, `upgrader.exe` cannot update it.

-> A contradiction!

Consider the following code:

```py
import subprocess

p = subprocess.Popen(['python.exe', '-c', 'import time; time.sleep(3); print("awaken")'])
print('Exiting')
```

Output:

```
PS D:\cli\testproj> python.exe main.py
Exiting
PS D:\cli\testproj> awaken
|
```

The first `python.exe` will exits and gives control back to the terminal. The second `python.exe` can print, yes, but it is out of order.

How to solve it? Wait for the subprocess to exit:

```py
import subprocess

p = subprocess.Popen(['python.exe', '-c', 'import time; time.sleep(3); print("awaken")'])
p.wait()
print('Exiting')
```

Output:

```
PS D:\cli\testproj> python.exe main.py
awaken
Exiting
PS D:\cli\testproj> |
```

Reply:
Yes, exactly.

----- 
I come up an idea to break the python dependency of `python.exe(1) -> upgrader.exe`.
We can modify `az.cmd`, if the param is `upgrade`, call upgrader.exe, otherwise, run az.

Furthermore, we can write the upgrade logic in bat, it looks feasible.